snapshot.go

```diff
@@ -56,19 +75,52 @@ func (s *SnapshotHandler) Save(a Aggregate) error {
 	return s.snapshotStore.Save(snap)
 }
 
-// Get fetch a snapshot and reconstruct an aggregate
-func (s *SnapshotHandler) Get(id string, a Aggregate) error {
-	typ := reflect.TypeOf(a).Elem().Name()
-	snap, err := s.snapshotStore.Get(id, typ)
+func (s *SnapshotHandler) saveAggregate(sa Aggregate) error {
+	root := sa.Root()
+	err := validate(*root)
 	if err != nil {
 		return err
 	}
-	err = s.serializer.Unmarshal(snap.State, a)
+	typ := reflect.TypeOf(sa).Elem().Name()
+	b, err := s.serializer.Marshal(sa)
 	if err != nil {
 		return err
 	}
-	root := a.Root()
-	root.setInternals(snap.ID, snap.Version, snap.GlobalVersion)
+	snap := Snapshot{
+		ID:            root.ID(),
+		Type:          typ,
+		Version:       root.Version(),
+		GlobalVersion: root.GlobalVersion(),
+		State:         b,
+	}
+	return s.snapshotStore.Save(snap)
+}
+
+// Get fetch a snapshot and reconstruct an aggregate
+func (s *SnapshotHandler) Get(id string, i interface{}) error {
+	typ := reflect.TypeOf(i).Elem().Name()
+	snap, err := s.snapshotStore.Get(id, typ)
+	if err != nil {
+		return err
+	}
+	switch a := i.(type) {
+	case SnapshotAggregate:
+		err := a.UnMarshal(s.serializer.Unmarshal, snap.State)
+		if err != nil {
+			return err
+		}
+		root := a.Root()
+		root.setInternals(snap.ID, snap.Version, snap.GlobalVersion)
```

Comment:
I'm not sure how this works by first calling Unmarshal on `a` followed by setInternals() on `root`. Would it be possible or beneficial to instead return a type instead of mutating values?

Reply:
I'm not sure I'm following on which level you mean when you say return a type?

Reply:
But yes there is some pointer magic going on. Not sure how to avoid it but up to suggestions ;)

Reply:
It's probably me that doesn't fully understand how to use this method.
But let me elaborate anyhow. If there's a method, e.g. Get in this case, that takes an ID to do a lookup something it is nice if the signature is something like:

 ```go
func Get(id string) (type, error)
```

But this is maybe not possible in this case because the type may either by Aggregate or SnapshotAggregate.

Reply:
>But this is maybe not possible in this case because the type may either by Aggregate or SnapshotAggregate.

Exactly the type has to come from the outside for the Get internal to populate it with data. Like the Marshal/Unmarshal pattern.